### ISSUE TITLE
resizing seems really slow

### ISSUE THREAD

**Issue by unphased** [jwilm/alacritty#969]
Which operating system does the issue occur on? macOS 10.13.2

Very sluggish resizing. hopefully any expensive render buffer resizing can occur in a debounced way at the end of the window resize.

**Comment by jwilm** [comment]
Can you elaborate on slow? The window size changes shape pretty smoothly on my laptop, but we don't actually draw during resize on macOS as of today. I think there might be a bug for that somewhere.

**Comment by jedahan** [comment]
I noticed this as well 
![slow-resize](https://user-images.githubusercontent.com/32407/34361388-ac28dd9c-ea37-11e7-968b-4766b747f222.gif)

**Comment by jwilm** [comment]
oh wow that's wild. I'm on an older version of macOS. Are you on high sierra as well, @jedahan ?

**Comment by unphased** [comment]
Yeah by slow we do mean ~2 seconds for it to eventually re-paint, and it seems to chug all while resizing the window also. It used to be a bit less slow last week, I would say.

**Comment by jedahan** [comment]
yeah high sierra, but I forgot how long I've noticed this for, its been a long while. I can try a git bisect especially if you give a decently recent starting point....

**Comment by jedahan** [comment]
Small update... this is bisect-able, almost done bisecting

**Comment by jedahan** [comment]
Hmm bisect says first bad commit is 5009566ea5c46a85fa243b18ce4b7fe8e0b89b62 which seems weird, gonna investigate more.

**Comment by jedahan** [comment]
@unphased can you try checking out 4e9b1c5 and see if the issue is gone?

Maybe it has to do with a similar bug as described in https://github.com/emoon/rust_minifb/issues/28 ?

No idea really...

**Comment by jedahan** [comment]
@unphased actually, even easier. This fixes resizing for me:

```diff
diff --git a/src/window.rs b/src/window.rs
index aec4b76..69c54a9 100644
--- a/src/window.rs
+++ b/src/window.rs
@@ -196,7 +196,7 @@ impl Window {
         Window::platform_window_init();
         let window = WindowBuilder::new()
             .with_title(title)
-            .with_transparency(true)
+            .with_transparency(false)
             .with_decorations(window_config.decorations());
         let context = ContextBuilder::new()
             .with_vsync(true);
```

**Comment by unphased** [comment]
Thank you, I will try this now.

**Comment by unphased** [comment]
Oh yeah. It's way faster. So probably something about how transparency is implemented is making it really sluggish. 

What should we do, though? This breaks transparency. I'm happy to give it up. Maybe we can add a config toggle for this.

**Comment by unphased** [comment]
We already know this but I can confirm that transparency is completely fine for performance on El Capitan

**Comment by jedahan** [comment]
From https://lists.apple.com/archives/quartz-dev/2008/Feb/msg00049.html

>It looks like we found the problem. Our Transform matrix had non- integer values in it for the translations. Obviously the Transform matrix is floating point, but if the translation values (tx & ty) were non-integer - 23.467 instead of 23.0 - then this put our drawing down the slow path and caused the resampling. If we round our translation values then our paints go down to 3-4 ms which is excellent.

So maybe we can have fast transparency if we round things instead of resample? Not sure what would need to be rounded. I'm gonna test 'native' scaled (2x) res with transparency and see if thats fast.

**Comment by jedahan** [comment]
Yeah that didn't help.

Whats weird is native fullscreen the transition is smooth

**Comment by unphased** [comment]
The fullscreen mode does not apply transparency, that is hopefully why!

**Comment by A1Liu** [comment]
I have this bug too, and found that when I compile from source with `cargo run --release` the problem goes away, but if I copy the binary from `target/release/alacritty` into the application folder with `cp target/release/alacritty /Applications/Alacritty.app/Contents/MacOS/alacritty` and then execute `/Applications/Alacritty.app/Contents/MacOS/alacritty`, it's slow again. Editing the source like @jedahan described and recompiling fixes the problem for me in both cases. What would need to be done to introduce a config option to disable `with_transparency`?

I'm running MacOS Mojave 10.14.5.

**Comment by chrisduerr** [comment]
I can imagine that #2438 might fix this. Though currently this is not available on macOS yet because some missing upstream features.

**Comment by sohailsomani** [comment]
Have this problem on Ubuntu as well, using latest.

**Comment by chrisduerr** [comment]
That sounds like it's tmux being slow. Make sure the same behavior isn't present in other terminals too.

**Comment by sohailsomani** [comment]
> That sounds like it's tmux being slow. Make sure the same behavior isn't present in other terminals too.

Good suggestion. I've been using gnome terminal and resizing never takes as long (seconds in ala vs milliseconds in gnome) though it isn't immediate. Is it possible that alacritty sends more resize messages than other terminals? I will try to debug it today.

**Comment by chrisduerr** [comment]
Seconds sounds extremely strange. Though if you need more help please open a separate issue since it's not related to this one.

**Comment by sohailsomani** [comment]
> Seconds sounds extremely strange. Though if you need more help please open a separate issue since it's not related to this one.

As you suggested, it seems tmux is to blame here. It is longer in alacritty (about 1-2 seconds), but there is a noticeable resize delay in gnome-terminal (about 1 second). Fair to put the blame on tmux here IMO.

**Comment by seeruk** [comment]
Just to keep this on up-to-date, I'm still experiencing this issue on MacOS 10.15.6 (with the supplemental update) on Alacritty 0.5.0. The aforementioned PR that was suspected to fix it had been merged, and it doesn't seem to have done the trick unfortunately. I'm still just going with compiling it transparency disabled for the time being - at which point I'm not experiencing any issues.

**Comment by ed-velez** [comment]
I want to say that this issue is still present. It may not be quite as exaggerated as it is on the video that was posted in this thread - but it's still present. I particularly notice it when I go to mission control and the Alacritty window is the window that last had focus. 

Additionally I found that if I just set `decorations: none` in my config and restart Alacritty I no longer see this issue. I suspect this is similar to trying to disable transparency directly in code.

**Comment by angelus2014** [comment]
I'm seeing this as well, on macOS 10.15.7, Mac Mini 2018 i3 with 16Gb. of RAM. Very sluggish when resizing the window. Settings 'decorations: none' doesn't help. I've compared Alacritty to the default Apple terminal, iTerm2 and Kitty and none of them display this behaviour, only Alacritty.


---
With regards,

Angelo Machils

**Comment by jef** [comment]
FWIW, I can't reproduce this. macOS 10.15.7, i9, 16GB RAM.

**Comment by akovardakov** [comment]
Same here.
Also noticed that font size also changes back and forth when I try to resize window with cursor.
macOS 10.15.7, MacBook Pro (16-inch, 2019) i7, 32GB RAM.

**Comment by seeruk** [comment]
> FWIW, I can't reproduce this. macOS 10.15.7, MacBook Pro (15-inch, 2019) i9, 16GB RAM.

Which GPU are you using? I don't have this problem if I'm plugged into an external monitor because it uses the dedicated GPU in my 16" MBP. If I unplug and use the Intel integrated GPU then I get the slow resizing.

**Comment by akovardakov** [comment]
> Which GPU are you using? I don't have this problem if I'm plugged into an external monitor because it uses the dedicated GPU in my 16" MBP. If I unplug and use the Intel integrated GPU then I get the slow resizing. (MacOS 10.15.7, Alacritty 0.6.0)
Confirm. It uses 100% of Intel graphics during resize. 
Is there a way to disable #210 ?

**Comment by seeruk** [comment]
Yeah, that's not an acceptable solution sadly. You can go from 10hr battery life to 2hr :(

**Comment by jef** [comment]
> If I unplug and use the Intel integrated GPU then I get the slow resizing.

> Edit: I disabled "Automatic graphic switching" in system preference and now lag is gone. But also every app uses only AMD now :)

Aha! This makes more sense now. I have it always using the AMD card 😄

**Comment by AndreiSva** [comment]
This is happening to me on arch linux.

**Comment by ngmmartins** [comment]
Installed this terminal today and I can also see it being slow when resizing. Also the text size also changes during the resize process.
Made a little video showcasing Alacritty, iTerm2 and the default Terminal.

https://user-images.githubusercontent.com/13188231/112498455-73047600-8d7e-11eb-92a9-dfe9b78255ff.mov

**Comment by jjant** [comment]
Any updates on this? Still really slow on Monterey.
